l1-contracts/contracts/upgrades/GatewayUpgrade.sol

```diff
@@ -0,0 +1,57 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity 0.8.24;
+
+import {Initializable} from "@openzeppelin/contracts-upgradeable/proxy/utils/Initializable.sol";
+
+import {BaseZkSyncUpgrade, ProposedUpgrade} from "./BaseZkSyncUpgrade.sol";
+
+import {DataEncoding} from "../common/libraries/DataEncoding.sol";
+
+import {Diamond} from "../state-transition/libraries/Diamond.sol";
+import {PriorityQueue} from "../state-transition/libraries/PriorityQueue.sol";
+import {PriorityTree} from "../state-transition/libraries/PriorityTree.sol";
+
+import {IGatewayUpgrade} from "./IGatewayUpgrade.sol";
+import {IL1SharedBridgeLegacy} from "../bridge/interfaces/IL1SharedBridgeLegacy.sol";
+/// @author Matter Labs
+/// @custom:security-contact security@matterlabs.dev
+/// @notice This upgrade will be used to migrate Era to be part of the hyperchain ecosystem contracts.
+contract GatewayUpgrade is BaseZkSyncUpgrade, Initializable {
+    using PriorityQueue for PriorityQueue.Queue;
+    using PriorityTree for PriorityTree.Tree;
+
+    /// @notice The main function that will be called by the upgrade proxy.
+    /// @param _proposedUpgrade The upgrade to be executed.
+    function upgrade(ProposedUpgrade calldata _proposedUpgrade) public override returns (bytes32) {
+        (address gatewayUpgradeAddress, bytes memory l2TxDataStart, bytes memory l2TxDataFinish) = abi.decode(
```

Comment:
to be even more precise, we could make the `gatewayUpgradeAddress` an immutable in this contract that is set in the constructor:

```
constructor() {
  THIS_ADDRESS = address(this);
}
```

this way less possibility of an error during upgrade calldata composition

Reply:
This should be in every contract tbh